test/python/apps/exchange.py

```diff
@@ -127,8 +88,7 @@ def __init__(self,
             self.subcommand_specs = FUND_SPECS
 
         self._exchange_partner = ExchangePartnerIdentity(self.subcommand_specs.curve, name)
```

Comment:
This is starting to look a bit intricate and not very clear on what is manipulated (`use_main_partner`, `use_test_key` booleans and other arguments).

I think the code was written with a **functional** behavior in mind: the ledger and partner keys were generated once and filled at the right place all the time. It was easier from a user point of view. But now we want to check for errors when using wrong keys, so the previous structure feels not appropriate anymore, we need more flexibility.

The crypto capacities could be removed from `ExchangeClient` and the test explicitely manage (instantiated and passed as `ExchangeClient::methods` argument) which key is used when - rather than modifying internal behaviors with boolean. I feel like it would be clearer:

```python
valid_ledger_signer = LedgerTestSigner(value=LEDGER_TEST_PRIVATE_KEY_INT)
# fake_ledger_signer = LedgerTestSigner()

first_partner_key = ExchangePartnerIdentity(curve, name)
second_partner_key = ExchangePartnerIdentity(curve, name)

client.set_partner_key(first_partner_key)

for expected_status, partner_key in [(9000, first_partner_key),
                                     (SIGN_VERIFICATION_FAIL, second_partner_key)] :
    rapdu = client.check_transaction(valid_ledger_signer, partner_key)
    assert rapdu.status == expected_status

[...]
```

Reply:
OK I managed to extract all signing authorities related code from the exchange app module, please let me know what you think of it !